### PR TITLE
[Auth] Support for providing custom Oauth 2.0 provider

### PR DESCRIPTION
**Pull request by ivarconr**
The intention of this PR is to solve the following issues:
- [x] [Auth] User-provider ( #261)
- [x] [Auth] Document how to secure Unleash (#234)
- [x] [Auth] Admin UI integrated with Oauth (#233)
- [x] [Auth] Admin UI should handle 401 (#232)
- [x] [Auth] Client API authentication (#231)
- [x] [Auth] Handle  `403 (Forbidden)` with custom auth.
- [x] [Auth] Support sign out (#288)

This feature also requires changes in the `unleash-frontend` module, code can be tracked in this PR:
https://github.com/Unleash/unleash-frontend/pull/99

Technical stuff that must be done before this is merged:
- [x]  clean up commits
- [x]  write tests: 
  - [x] for "simple-authentication"
  - [x] for "custom-authentication" 

I have written a small example on how you would provide and Google Oauth 2.0 integration via passport here: 
https://gist.github.com/ivarconr/3748f20e5c5c0634879a2b01b957911e

### CODE REVIEW

**Comment by ivarconr** [comment]
Should also test this approach with basic authentication

File: lib/app.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ const unleashSession = require('./middleware/session');
 const responseTime = require('./middleware/response-time');
 const requestLogger = require('./middleware/request-logger');
 const validator = require('./middleware/validator');
+const authBuiltin = require('./auth-builtin');
```
Comment (sveisvei) [inline lib/app.js]: Not sure I feel the name is awesoem, if we stick with it though - it should be `built-in`

File: lib/user.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+const gravatar = require('gravatar');
+
+module.exports = class User {
+    constructor({ name, email, imageUrl }) {
+        this.email = email;
```
Comment (sveisvei) [inline lib/user.js]: Missing validation of email.

File: test/e2e/api/admin/feature.e2e.test.js [hunk 3]
```diff
@@ -50,7 +50,7 @@ test.serial('creates new feature toggle', async t => {
         .then(destroy);
 });
 
-test.serial('creates new feature toggle with createdBy', async t => {
+test.skip.serial('creates new feature toggle with createdBy', async t => {
```
Comment (sveisvei) [inline test/e2e/api/admin/feature.e2e.test.js]: Intended?

File: lib/app.js (hunk 1 above)
Comment (sveisvei) [reply lib/app.js]: Alternatives `simple-session` / `simple-cookie`  or something.....

Comment (ivarconr) [reply lib/app.js]: agreed, we need a better name!

File: lib/user.js (hunk 2 above)
Comment (ivarconr) [reply lib/user.js]: :+1:

File: test/e2e/api/admin/feature.e2e.test.js (hunk 3 above)
Comment (ivarconr) [reply test/e2e/api/admin/feature.e2e.test.js]: No, must be fixed. Have not focused on tests yet. Need to ramp my TDD skill i suppose.

File: lib/user.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+        this.name = name;
```
Comment (sveisvei) [inline lib/user.js]: Missing validation on name.

File: lib/auth-builtin.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const User = require('./user');
+const AuthenticationRequired = require('./authentication-required');
+
+function enableUnsecureAuth(app) {
+    app.post('/api/admin/login', (req, res) => {
+        const user = req.body;
+        req.session.user = new User({ email: user.email });
```
Comment (sveisvei) [inline lib/auth-builtin.js]: `User` takes name/uri as well, support it?

File: test/e2e/api/admin/feature.e2e.test.js (hunk 3 above)
Comment (sveisvei) [reply test/e2e/api/admin/feature.e2e.test.js]: No, TDD skills are so 2012. 🎉

File: lib/app.js (hunk 1 above)
Comment (ivarconr) [reply lib/app.js]: how about `unsecure-session`  or `unsecure-authentication`?

File: lib/user.js (hunk 4 above)
Comment (ivarconr) [reply lib/user.js]: The idea was to make it optional. It feels a bit overkill to have the user specify `name` and `email` when using the builtin cookie based session. We are only using the email to build gravatar and tag changes.  

What do you think?

Comment (sveisvei) [reply lib/user.js]: Indeed optional, but could pick it up anyhow. Currently the builtin does not pass it along.

Suggestion: We could also default the name from the email, e.g.`/^(.*)@/i`

File: lib/middleware/simple-authentication.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            next();
+        } else {
+            return res
+                .status('401')
+                .json(
+                    new AuthenticationRequired({
+                        path: '/api/admin/login',
+                        type: 'unsecure',
+                        message:
+                            'You have to indetify yourself in order to use Unleash.',
```
Comment (ivarconr) [inline lib/middleware/simple-authentication.js]: typo: indetify => identify

File: package.json [hunk 7]
```diff
@@ -56,6 +56,7 @@
     ]
   },
   "dependencies": {
+    "assert": "^1.4.1",
```
Comment (SimenB) [inline package.json]: assert is bundled with node, this shouldn't be here

File: lib/user.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const gravatar = require('gravatar');
+const assert = require('assert');
```
Comment (SimenB) [inline lib/user.js]: why not use Joi for this?

Comment (ivarconr) [reply lib/user.js]: Good question.

**Comment by panel** [comment]
Is there anything I can do to help on this PR? My organization is looking at Unleash as an option for feature flag management, but authentication, authorization groups and audit logs are a big requirement. I have some bandwidth to help push this forward.

**Comment by ivarconr** [comment]
Hi Paul! 

Thanks for reaching out. This PR will at least solve your authentication and audit log requirement, and make it possible to add the auth-provider you need. I will merge this to master shortly and have an alpha-release out that you can play with. I want multiple users of unleash to play with the alpha release for some time before making an official v3. release of unleash. 

When it comes to your authorization groups requirements it would be nice if you could describe your needs in a separate issue request and we can take the discussions from there.

**Comment by panel** [comment]
@ivarconr Just wanted to give you a shoutout. I was able to use the `preHook` and `preRouterHook` to hook into our OAuth2 identity provider in a mater of minutes! The hooks provided are simple and powerful, and the example snippets were helpful.

The only bit that was surprising/took some digging in the source was the ability to return a 401 with a `message` and `path` parameter to prompt the user to login. That flow was way easier than I expected, but I sort of stumbled into it and figured it out by jumping into the client source. I'm planning on opening a PR to add to the auth documentation based on my experience.

**Pull request merged**